nxp/s32k3: add C40 internal flash driver + MCUboot flow for MR-CANHUBK3

sumitbatra-nxp opened the pull request:
This series brings full internal-flash (C40) support and an end-to-end
MCUboot flow to the NXP MR-CANHUBK3 (S32K344) board.

**What’s included**

**New driver:** drivers/flash/flash_mcux_c40.c for the S32K3x C40 array

8 KiB erase / 8-byte write unit, page layout reporting

D-cache invalidation after mutating ops

Optional protection policy to lock IVT/MCUboot partitions at init
(CONFIG_SOC_FLASH_MCUX_C40_APPLY_PROTECTION)

Automatic relocation of driver + HAL to SRAM when XIP

**Bindings & SoC DTSI**: nxp,s32k3x-c40-flash.yaml and flash0 node in
nxp_s32k344_m7.dtsi (left status = "disabled" by default)

**IVT/linker gating:** place .ivt_header only when XIP and either
standalone app or MCUboot itself (avoid clashing with MCUboot image header
when chainloaded)

**Board support:** MCUboot partition layout & code-partition overlays for
mr_canhubk3, plus a small doc update

**BUILD MCUBOOT AND FLASH_SHELL APP FOR CHAINLOAD**
west build -p -b mr_canhubk3/s32k344/mcuboot --sysbuild   -s zephyr/samples/drivers/flash_shell   -d build/sys_mcuboot   --     -DSB_CONFIG_BOOT_SIGNATURE_TYPE_RSA=y   -DSB_CONFIG_BOOT_SIGNATURE_KEY_FILE=\"${PWD}/bootloader/mcuboot/root-rsa-2048.pem\"

**These changes need NXP_HAL PR -** 
https://github.com/zephyrproject-rtos/hal_nxp/pull/618

**For now I have disabled the flash0 node, till the time HAL changes are merged.**

sumitbatra-nxp commented on boards/nxp/mr_canhubk3/doc/index.rst:
```diff
@@ -296,6 +296,81 @@ For example, to erase and verify flash content:
+
+   west build -b mr_canhubk3 -d build/mcuboot bootloader/mcuboot/boot/zephyr \
+     -DEXTRA_DTC_OVERLAY_FILE="boards/nxp/mr_canhubk3/mr_canhubk3_mcuboot_layout.overlay;boards/nxp/mr_canhubk3/mr_canhubk3_mcuboot_boot.overlay"
+
```
please check now

sumitbatra-nxp commented on boards/nxp/mr_canhubk3/mr_canhubk3_mcuboot_app.overlay:
please check now

JarmouniA commented:
@sumitbatra-nxp Please leave comments' resolving to reviewers.

JarmouniA commented on dts/bindings/flash_controller/nxp,s32k3x-c40-flash.yaml:
```diff
@@ -0,0 +1,19 @@
+compatible: "nxp,s32k3x-c40-flash"
+
+# Inherit common SoC flash schema (reg/status, partitions, etc.)
+include: soc-nv-flash.yaml
```
The flash controller is distinct from the Flash chip, this compatible should be in the Flash chip node.

sumitbatra-nxp replied:
Moved it to mtd

JarmouniA replied:
> Moved it to mtd

Why!?

I was referring to the inclusion of `soc-nv-flash.yaml` here, in a flash controller binding. Is this flash controller used with one type of flash? So we can assume erase-block-size & write-block-size are always the same.

JarmouniA replied:
> The NXP SoC's that I have come across so far, this flash controller is used with this type of flash only.

Then define a binding for the Flash itself as well, and move erase-block-size & write-block-size there.

sumitbatra-nxp commented on drivers/flash/CMakeLists.txt:
```diff
@@ -167,6 +167,16 @@ if(CONFIG_FLASH_NXP_S32_QSPI_NOR OR CONFIG_FLASH_NXP_S32_QSPI_HYPERFLASH)
+if(CONFIG_FLASH_MCUX_C40_API AND CONFIG_CODE_DATA_RELOCATION)
+  zephyr_code_relocate(FILES ${CMAKE_CURRENT_LIST_DIR}/flash_mcux_c40.c LOCATION RAM)
+  if(DEFINED ZEPHYR_HAL_NXP_MODULE_DIR)
+      # Relocating HAL driver from Zephyr keeping HAL RTOS-agnostic
```
Corrected !

sumitbatra-nxp commented on west.yml:
```diff
@@ -210,7 +210,7 @@ manifest:
       groups:
         - hal
     - name: hal_nxp
-      revision: 4377ecfba52fe0ff7352eadf426b523ed3e1d27f
```
Corrected !

sumitbatra-nxp commented on boards/nxp/mr_canhubk3/Kconfig.mr_canhubk3_s32k344_mcuboot:
Removed as suggested !

manuargue commented on drivers/flash/Kconfig.mcux:
```diff
@@ -130,3 +130,32 @@ config FLASH_MCUX_XSPI
+	imply CODE_DATA_RELOCATION_SRAM if XIP
+	help
+	  Enable the MCUX C40 internal flash driver used on NXP S32K3x
+	  (e.g. S32K344). Provides Zephyr flash driver using MCUX C40 HAL.
```
I would remove references to S32K3, this driver can be used for other NXP SoCs featuring C40.

manuargue commented on drivers/flash/Kconfig.mcux:
```diff
@@ -130,3 +130,32 @@ config FLASH_MCUX_XSPI
 	help
 	  Enable the mcux xspi flash driver.
+
+if DT_HAS_NXP_S32K3X_C40_ENABLED
```
remove and use instead `depends on DT_HAS_NXP_C40_ENABLED`. Note that I intentionally dropped the `S32K3X` from the name beacuse we should aim to have a single DT binding for NXP C40.

manuargue commented on drivers/flash/Kconfig.mcux:
```diff
@@ -130,3 +130,32 @@ config FLASH_MCUX_XSPI
+	  (e.g. S32K344). Provides Zephyr flash driver using MCUX C40 HAL.
+
+# NXP MCUX C40 internal flash: optional protection pass at init
+config SOC_FLASH_MCUX_C40_APPLY_PROTECTION
```
guard with `if SOC_FLASH_MCUX_C40` and drop `depends on SOC_FLASH_MCUX_C40` on this config. See how `config SOC_FLASH_MCUX` was done above, that's the correct approach.

manuargue commented on dts/bindings/mtd/nxp,s32k3x-c40.yaml:
```diff
@@ -0,0 +1,19 @@
+description: |
+  On-chip C40 flash used in NXP S32K3x MCUs (e.g. S32K344).
+
+compatible: "nxp,s32k3x-c40"
```
```diff
- compatible: "nxp,s32k3x-c40"
+ compatible: "nxp,c40"
```
there's nothing specific to s32k3, we can reuse this binding for other SoCs featuring C40 flash.

JarmouniA replied:
`nxp,c40-flash`

manuargue commented on dts/bindings/mtd/nxp,s32k3x-c40.yaml:
```diff
@@ -0,0 +1,19 @@
+# SPDX-License-Identifier: Apache-2.0
+title: NXP S32K3x C40 Internal Flash Controller
```
```diff
- title: NXP S32K3x C40 Internal Flash Controller
+ title: NXP C40 Internal Flash Controller
```

manuargue commented on dts/arm/nxp/nxp_s32k344_m7.dtsi:
```diff
@@ -70,7 +70,10 @@
 		 * created, based on it.
```
pls remove this comment, your changes render it obsolete:
> Do not assign the compatible for this now, when Flash API is implemented,
> need to check if "soc-nv-flash" can be used or a new binding need to be
> created, based on it.

manuargue reviewed (changes requested):
@sumitbatra-nxp pls ensure the CI builds are green. There are currently some errors in compliance, docs and twister.

JarmouniA commented on boards/nxp/mr_canhubk3/doc/index.rst:
```diff
@@ -296,6 +296,75 @@ For example, to erase and verify flash content:
+-d build/sys_mcuboot
+-- 
+-Dmcuboot_CONFIG_BUILD_OUTPUT_HEX=y
+-Dflash_shell_CONFIG_BUILD_OUTPUT_HEX=y
```
```suggestion
   $ west build -p -b mr_canhubk3/s32k344/mcuboot samples/drivers/flash_shell --sysbuild
```

JarmouniA commented on boards/nxp/mr_canhubk3/doc/index.rst:
```diff
@@ -296,6 +296,75 @@ For example, to erase and verify flash content:
 
    west flash -r trace32 --startup-args elfFile=build/zephyr/zephyr.elf loadTo=flash eraseFlash=yes verifyFlash=yes
 
+MCUboot (sysbuild)
```
```suggestion
MCUboot
```

JarmouniA commented on boards/nxp/mr_canhubk3/mr_canhubk3_s32k344_mcuboot.dts:
```diff
@@ -0,0 +1,62 @@
+		};
+
+		/* Scratch (swap) area */
+		scratch_partition: partition@3A2000 {
```
not needed, unless you want to use swap-using-scratch (not recommended)

JarmouniA commented on boards/nxp/mr_canhubk3/mr_canhubk3_s32k344_mcuboot_defconfig:
```diff
@@ -0,0 +1,6 @@
+
+CONFIG_BOOTLOADER_MCUBOOT=y
+# S32K3 vector table forces a 1 KiB MCUboot header
+CONFIG_ROM_START_OFFSET=0x400
```
Should probably be set at the SoC level. (non-blocking)

JarmouniA commented on drivers/flash/CMakeLists.txt:
```diff
@@ -184,6 +184,16 @@ if(CONFIG_FLASH_NXP_S32_QSPI_NOR OR CONFIG_FLASH_NXP_S32_QSPI_HYPERFLASH)
+if(CONFIG_SOC_FLASH_MCUX_C40 AND CONFIG_CODE_DATA_RELOCATION)
+  zephyr_code_relocate(FILES ${CMAKE_CURRENT_LIST_DIR}/flash_mcux_c40.c LOCATION RAM)
+  if(DEFINED ZEPHYR_HAL_NXP_MODULE_DIR)
+    # Relocating HAL driver from Zephyr keeping HAL RTOS-agnostic
```
What does this mean?

JarmouniA commented on dts/arm/nxp/nxp_s32k344_m7.dtsi:
```diff
@@ -70,7 +70,10 @@
+			compatible = "nxp,s32k3x-c40";
 			reg = <0x00400000 DT_SIZE_K(4048)>;
+			erase-block-size = <DT_SIZE_K(8)>;     /* Min erase size is 8 KiB */
+			write-block-size = <8>;                /* Minimum programmable unit */
```
is this in bytes?

JarmouniA commented on soc/nxp/s32/s32k3/soc.c:
```diff
@@ -33,21 +38,24 @@ struct ivt {
 };
-#endif /* CONFIG_XIP */
+#endif
+#endif
```
```suggestion
#endif /* CONFIG_XIP */
```

sumitbatra-nxp commented on boards/nxp/mr_canhubk3/mr_canhubk3_s32k344_mcuboot_defconfig:
```diff
@@ -0,0 +1,6 @@
+
+CONFIG_BOOTLOADER_MCUBOOT=y
+# S32K3 vector table forces a 1 KiB MCUboot header
+CONFIG_ROM_START_OFFSET=0x400
```
If I remove CONFIG_ROM_START_OFFSET=0x400 from here then mcuboot image considers it as 0 whereas flash_shell app has the correct value 0x400

grep ROM_START_OFFSET build/mcuboot/zephyr/.config
CONFIG_ROM_START_OFFSET=0
grep ROM_START_OFFSET build/flash_shell/zephyr/.config
CONFIG_ROM_START_OFFSET=0x400

Dat-NguyenDuy commented on drivers/flash/CMakeLists.txt:
```diff
@@ -184,6 +184,17 @@ if(CONFIG_FLASH_NXP_S32_QSPI_NOR OR CONFIG_FLASH_NXP_S32_QSPI_HYPERFLASH)
+
+if(CONFIG_SOC_FLASH_MCUX_C40 AND CONFIG_CODE_DATA_RELOCATION)
+  # Relocate the Zephyr driver to RAM when relocation is enabled
+  zephyr_code_relocate(FILES ${CMAKE_CURRENT_LIST_DIR}/flash_mcux_c40.c LOCATION RAM)
```
does it work?

nordicjm commented on drivers/flash/Kconfig.mcux:
```diff
@@ -130,3 +130,32 @@ config FLASH_MCUX_XSPI
+if DT_HAS_NXP_C40_FLASH_ENABLED
+# MCUX C40 internal flash driver
+config SOC_FLASH_MCUX_C40
+	def_bool y
```
```suggestion
	bool "MCUX C40 flash driver"
	default y
	depends on DT_HAS_NXP_C40_FLASH_ENABLED
```
or similar

nordicjm commented on drivers/flash/Kconfig.mcux:
```diff
@@ -130,3 +130,32 @@ config FLASH_MCUX_XSPI
 	  Enable the mcux xspi flash driver.
+
+if DT_HAS_NXP_C40_FLASH_ENABLED
+# MCUX C40 internal flash driver
```
```suggestion
```

nordicjm commented on drivers/flash/flash_mcux_c40.c:
```diff
@@ -0,0 +1,377 @@
+	struct prot_range prot_aligned[8];
+	size_t nprot_al = 0;
+	size_t i;
+	uint32_t s, e, sa, ea;
```
one declaration per line, https://kernel.org/doc/html/latest/process/coding-style.html

> To this end, use just one data declaration per line (no commas for multiple data declarations).

nordicjm commented on boards/nxp/mr_canhubk3/doc/index.rst:
```diff
@@ -296,6 +296,58 @@ For example, to erase and verify flash content:
+
+.. code-block:: console
+
+ $west build -p -b mr_canhubk3/s32k344/mcuboot samples/drivers/flash_shell --sysbuild
```
use the proper west build doxygen command for this which will format it nicely

sumitbatra-nxp replied:
Please check now

nordicjm commented on boards/nxp/mr_canhubk3/doc/index.rst:
```diff
@@ -296,6 +296,58 @@ For example, to erase and verify flash content:
+
+ $ west build -p -b mr_canhubk3/s32k344/mcuboot samples/drivers/flash_shell --sysbuild \
+     -- -DSB_CONFIG_BOOT_SIGNATURE_TYPE_RSA=y \
+     -DSB_CONFIG_BOOT_SIGNATURE_KEY_FILE="$PWD/bootloader/mcuboot/root-rsa-2048.pem"
```
as above

nordicjm commented on soc/nxp/s32/s32k3/Kconfig.defconfig:
```diff
@@ -28,4 +28,9 @@ endif # ETH_DRIVER && NET_L2_ETHERNET
+# S32K3 places the vector table at +0x400
+# This makes the bootloader and the sign tool compatible with that.
+config ROM_START_OFFSET
+	default 0x400 if (BOOTLOADER_MCUBOOT || MCUBOOT)
```
then this would be for `BOOTLOADER_MCUBOOT` only, otherwise it will shift MCUboot itself by 0x400 and never boot it

sumitbatra-nxp replied:
@nordicjm  My MCUboot Vector Table is also at 0x400 offset (0x402400)
IVT's  4th word also points to 0x00402400

nordicjm replied:
So what is the first 0x400? Why doesn't the flash controller point to the start of usable flash?

sumitbatra-nxp replied:
@nordicjm .. Corrected.. please check now

JarmouniA commented on boards/nxp/mr_canhubk3/Kconfig.sysbuild:
```diff
@@ -0,0 +1,15 @@
+
+choice BOOT_SIGNATURE_TYPE
+    default BOOT_SIGNATURE_TYPE_RSA
+endchoice
```
Still not addressed.

sumitbatra-nxp replied:
Check now

JarmouniA commented on dts/bindings/flash_controller/nxp,c40-flash-controller.yaml:
```diff
@@ -0,0 +1,25 @@
+    const: 1
+    description: Number of size cells for child flash nodes.
+
+  ranges:
```
> not needed, defined in DT spec by default.

Still not addressed.

JarmouniA commented on soc/nxp/s32/s32k3/soc.c:
```diff
@@ -12,7 +12,12 @@
 #include <OsIf.h>
 
-#ifdef CONFIG_XIP
+#if defined(CONFIG_XIP)
```
> unnecessary change

Still not addressed.

JarmouniA commented on drivers/flash/flash_mcux_c40.c:
```diff
@@ -0,0 +1,377 @@
+ * Zephyr driver for NXP S32K3x C40 internal flash
+ */
+
+#define DT_DRV_COMPAT nxp_c40_flash
```
Get them from the child node, the driver should be for the controller (the actual peripheral) not the flash chip.

JarmouniA commented on boards/nxp/mr_canhubk3/mr_canhubk3_s32k344_mcuboot.dts:
```diff
@@ -0,0 +1,56 @@
+ */
+
+/dts-v1/;
+#include "mr_canhubk3.dts"
```
You can't include a dts file, only dtsi.
Make a common dtsi for both variants.

sumitbatra-nxp replied:
@JarmouniA ..Please check now

JarmouniA commented on drivers/flash/flash_mcux_c40.c:
```diff
@@ -0,0 +1,381 @@
+		return mcux_to_errno(st);
+	}
+
+	LOG_INF("C40 flash: base=0x%lx size=0x%lx erase=0x%lx write=0x%lx",
```
```suggestion
	LOG_DBG("C40 flash: base=0x%lx size=0x%lx erase=0x%lx write=0x%lx",
```
Apply for all occurrences.

sumitbatra-nxp commented on dts/arm/nxp/nxp_s32k344_m7.dtsi:
```diff
@@ -63,15 +63,24 @@
+				compatible = "nxp,c40-flash";
+				reg = <0x0 DT_SIZE_K(4048)>;
+				erase-block-size = <DT_SIZE_K(8)>;  /* 8 KiB sectors */
+				write-block-size = <8>;             /* 8-byte min program unit */
```
If I remove these from the SoC dtsi, I get symbol undefined build errors. The error goes away if I use default rather.
I also see this precedence of using const in yaml and still having the property defined in the dtsi in following files- 

Property: compensated-timer-freq
dts/bindings/timer/st,stm32wb0-radio-timer.yaml
dts/arm/st/wb0/stm32wb0.dtsi

sumitbatra-nxp commented on boards/nxp/mr_canhubk3/mr_canhubk3_s32k344_mcuboot.yaml:
```diff
@@ -0,0 +1,14 @@
+ram: 320
+flash: 1919
+vendor: nxp
+extends: mr_canhubk3
```
removed

sumitbatra-nxp commented on boards/nxp/mr_canhubk3/mr_canhubk3.dts:
Corrected.. please check now

sumitbatra-nxp commented:
@JarmouniA .. Twister gives me an error while building a test case in mcuboot  - undefined boot label 'boot_partition', so I am adding an additional boot_partition label  to my mcuboot partition. Like this -
boot_partition: mcuboot: partition@2000 {
                        label = "mcuboot";
                        reg = <0x00002000 DT_SIZE_K(128)>;
                };

I hope that is ok ?

nordicjm commented on boards/nxp/mr_canhubk3/doc/index.rst:
```diff
@@ -296,6 +296,74 @@ For example, to erase and verify flash content:
+   :board: mr_canhubk3/s32k344/mcuboot
+   :goals: build
+   :build-dir: build
+   :gen-args: --sysbuild -- -DSB_CONFIG_BOOT_SIGNATURE_TYPE_RSA=y -DSB_CONFIG_BOOT_SIGNATURE_KEY_FILE="$PWD/bootloader/mcuboot/root-rsa-2048.pem"
```
drop `-- -DSB_CONFIG_BOOT_SIGNATURE_TYPE_RSA=y -DSB_CONFIG_BOOT_SIGNATURE_KEY_FILE="$PWD/bootloader/mcuboot/root-rsa-2048.pem"` these are the defaults

manuargue commented:
All my comments addressed, @sumitbatra-nxp pls fix the CI builds to approve. Thanks

sumitbatra-nxp commented on drivers/flash/flash_mcux_c40.c:
```diff
@@ -0,0 +1,383 @@
+}
+
+struct prot_range {
+	uint32_t off, len;
```
corrected

JarmouniA commented on boards/nxp/mr_canhubk3/doc/index.rst:
```diff
@@ -296,6 +296,57 @@ For example, to erase and verify flash content:
+   :zephyr-app: samples/drivers/flash_shell
+   :board: mr_canhubk3/s32k344/mcuboot
+   :goals: build
+   :build-dir: build
```
```suggestion
```

sumitbatra-nxp replied:
removed

sumitbatra-nxp commented:
Manifest change reverted as requested. 
HAL PR is ready for merge.

@JarmouniA If you have the rights, can you please approve.

manuargue commented:
Hi @nordicjm , can you please check if your comments has been addressed? Thanks!

sumitbatra-nxp commented on drivers/flash/flash_mcux_c40.c:
```diff
@@ -0,0 +1,383 @@
+	return (a_off < b_end) && (b_off < a_end);
+}
+
+static int mcux_c40_read(const struct device *dev, off_t off, void *buf, size_t len)
```
Check now

sumitbatra-nxp commented on west.yml:
```diff
@@ -210,7 +210,7 @@ manifest:
         - hal
     - name: hal_nxp
-      revision: 6d6ab91d5cc71c872353af1957d19f3d930df0fd
+      revision: 59a3b6c6ec1aa4ed11c2a4ba75f197f3b3a9c43d
```
@dleach02 
Restored west.yml to the original hal_nxp revision and will wait for the HAL PR to merge before adding a manifest bump. 
Once the HAL is merged I’ll re-order commits so the manifest bump comes before the driver changes. 
Please let me know if there is still any gap in my understanding.

manuargue replied:
>Once the HAL is merged I’ll re-order commits so the manifest bump comes before the driver changes.

Do this even if the manifest is pointing to the unmerged hal_nxp pr. Then after the hal_nxp pr is merged, you only need to change in the manifest the pr to the merged sha. It will ease review.

sumitbatra-nxp replied:
@nordicjm Changed the sequence of commits as requested

sumitbatra-nxp committed:
drivers: flash: Add MCUX C40 flash controller driver

Add flash controller driver for the on-chip C40 flash controller
The driver is backed by the MCUX C40 HAL and implements
read/erase/program, page layout, and an optional protection policy
that can lock well-known regions (IVT/MCUboot) derived from devicetree.

Key details:
- Selects FLASH_HAS_DRIVER_ENABLED / FLASH_HAS_EXPLICIT_ERASE /
FLASH_HAS_PAGE_LAYOUT.
- Runs erase/program from SRAM when XIP by relocating both the shim and
the MCUX HAL source if CODE_DATA_RELOCATION_SRAM=y.
- Optional protection pass at init (FLASH_MCUX_C40_APPLY_PROTECTION),
which aligns windows to sector boundaries and applies lock/unlock
using the HAL. This is useful on XIP systems to keep IVT/bootloader
ranges read-only; can be disabled if a bootloader or security policy
manages protection instead.

Files:
- drivers/flash/flash_mcux_c40.c (new)
- drivers/flash/CMakeLists.txt (+zephyr_code_relocate when needed,
Compliance fixes)
- drivers/flash/Kconfig.mcux (enable flash driver, reloc & protection)
- modules/hal_nxp/mcux/mcux-sdk-ng/drivers/drivers.cmake

Signed-off-by: Sumit Batra <sumit.batra@nxp.com>

sumitbatra-nxp committed:
dts: nxp: Add binding for C40 flash and flash controller

Introduce DT bindings for on-chip C40 flash and its controller
and describe their corresponding nodes in nxp_s32k344_m7.dtsi.

- Binding: dts/bindings/mtd/nxp,c40-flash.yaml
Erase/write block sizes.

- Binding: dts/bindings/flash_controller/nxp,c40-flash-controller.yaml
Describe flash device (child) ranges

- SoC nodes:  With the new compatible and geometry
properties. Keep status = "disabled" at the SoC level
so boards opt-in.

This prepares the platform for using Zephyr’s flash API / FLASH_MAP /
MCUboot with internal code flash.

Signed-off-by: Sumit Batra <sumit.batra@nxp.com>

sumitbatra-nxp committed:
soc: nxp s32k3: derive sys clock from devicetree

Use devicetree to provide the system clock frequency for S32K3
instead of hardcoding it in board defconfigs.

- Add clock-frequency to /cpus/cpu@0 in nxp_s32k344_m7.dtsi using
DT_FREQ_M(160).
- Define DT_SYSCLK_PATH and derive SYS_CLOCK_HW_CYCLES_PER_SEC from
the sysclk node via dt_node_int_prop_int() when CORTEX_M_SYSTICK.
- Remove CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC from mr_canhubk3

This keeps the clock configuration in a single SoC-level place,
aligns S32K3 with other NXP Cortex-M SoCs, and ensures both the
MCUboot and application builds share the same
SYS_CLOCK_HW_CYCLES_PER_SEC.

Signed-off-by: Sumit Batra <sumit.batra@nxp.com>

sumitbatra-nxp committed:
soc: nxp: s32k3: gate IVT emission when chain-loaded by MCUboot

Emit the IVT section and IVT header only when XIP and the image is either
a standalone XIP app or MCUboot itself. Do not emit the IVT when the
Zephyr image is chain-loaded by MCUboot (BOOTLOADER_MCUBOOT=y).

- linker.ld/sections.ld: place .ivt_header at IVT_HEADER only under
XIP && (!BOOTLOADER_MCUBOOT || MCUBOOT).
Provide __ivt_region_start/end symbols.
- soc.c: guard IVT struct under the same condition and mark it 'used'
so the linker keeps it when needed.
- Kconfig.defconfig - Make the bootloader and the sign tool compatible
with the vector table.

This avoids populating 0x400000 IVT from the app image while retaining it
for MCUboot or standalone XIP use-cases.

Files:
- soc/nxp/s32/s32k3/linker.ld
- soc/nxp/s32/s32k3/sections.ld
- soc/nxp/s32/s32k3/soc.c
- soc/nxp/s32/s32k3/Kconfig.defconfig

Signed-off-by: Sumit Batra <sumit.batra@nxp.com>

sumitbatra-nxp committed:
boards: nxp: mr_canhubk3: add mcuboot variant and document sysbuild

Add an mcuboot board variant under the S32K344 SoC in board.yml.
New target is - mr_canhubk3/s32k344/mcuboot. This enables a
single sysbuild invocation to build MCUboot and the application.
Also make a common dtsi for all mr_canhubk3 board variants.

Extend the board documentation with an “MCUboot (sysbuild)” section
that:

Shows one-command sysbuild to produce both MCUboot and app
Documents signing via sysbuild (SB_CONFIG_BOOT_SIGNATURE_*),
and where the signed artifacts are generated.

Provides simple flashing instructions and a troubleshooting
note about using a 1 KiB (0x400) image header and not emitting
an IVT for chain-loaded apps.

Signed-off-by: Sumit Batra <sumit.batra@nxp.com>

kartben commented:
@sumitbatra-nxp this needs to be rebased please
